proj/src/main/webapp/script.js

```diff
@@ -27,17 +29,101 @@ if(btnLogin){
     });
 }
 
+
+// Adds user to an existing chat when given a reference to the place in the database
+function addUserToTag(reference){
+    var currentUID = firebase.auth().currentUser.uid
+    console.log("adding new user to chat room with uid: " + currentUID);
+    reference.push(currentUID);
+}
+
+// Creates a new chat given a tag and adds the current user as a member
+function createNewChatWithUser(tag){
+    console.log("creating new chat with tag: " + tag);
+    var newChat = {
+        "name" : tag,
+        "tag" : tag,
+    };
+    var currentReference = firebase.database().ref("/chat/" + tag);
+    var postKey = currentReference.push(newChat).key;
+    currentReference = firebase.database().ref("/chat/" + tag + "/" + postKey + "/users/");
+    addUserToTag(currentReference);
+
+    return postKey;
+}
+
+// Loops through open chat rooms, adds the user to the first open chat and returns the key
+function findChatAndAddUser(snapshot){
+    var key;
+    var foundOpenRoom = false;
+    snapshot.forEach(function(childSnapshot){
+        key = childSnapshot.key;
+        console.log("key1 " + key);
+
+        var userSnapshot = childSnapshot.child("users");
+        var usersReference = userSnapshot.ref;
+
+        // Handled below if spillover is needed
+        if(userSnapshot.numChildren() < MAX_CHAT_SIZE){
+            addUserToTag(usersReference);
+            foundOpenRoom = true;
+            return true;
```

Comment:
Maybe it is a task for a subsequent PR so this one doesn't get even more complicated, but what if the user already exists in a chatroom? We might want to check that we are not adding them twice or adding them to a second room. Also I just learned that a reference is an acceptable datatype. https://firebase.google.com/docs/firestore/manage-data/data-types
I wonder if you can do something like have the list of users just be a list of references to the user objects or maybe the users tag map will map tag to a reference to the chat or something. But again, that is a bit more complex and probably better as a follow up PR.